src/storage/mod.rs

```diff
@@ -392,61 +402,40 @@ impl Storage {
         let compressed_index_content =
             compress(BufReader::new(fs::File::open(&local_index_path)?), alg)?;
 
-        self.store_inner(
-            vec![
-                Blob {
-                    path: archive_path.to_string(),
-                    mime: "application/zip".to_owned(),
-                    content: zip_content,
-                    compression: None,
-                    date_updated: Utc::now(),
-                },
-                Blob {
-                    path: remote_index_path,
-                    mime: "application/octet-stream".to_owned(),
-                    content: compressed_index_content,
-                    compression: Some(alg),
-                    date_updated: Utc::now(),
-                },
-            ]
-            .into_iter()
-            .map(Ok),
-        )?;
+        self.store_inner(vec![
+            Blob {
+                path: archive_path.to_string(),
+                mime: "application/zip".to_owned(),
+                content: zip_content,
+                compression: None,
+                date_updated: Utc::now(),
+            },
+            Blob {
+                path: remote_index_path,
+                mime: "application/octet-stream".to_owned(),
+                content: compressed_index_content,
+                compression: Some(alg),
+                date_updated: Utc::now(),
+            },
+        ])
+        .await?;
 
         let file_alg = CompressionAlgorithm::Bzip2;
         Ok((file_paths, file_alg))
     }
 
-    fn transaction<T, F>(&self, f: F) -> Result<T>
-    where
-        F: FnOnce(&mut dyn StorageTransaction) -> Result<T>,
-    {
-        let mut conn;
-        let mut trans: Box<dyn StorageTransaction> = match &self.backend {
-            StorageBackend::Database(db) => {
-                conn = db.start_connection()?;
-                Box::new(conn.start_storage_transaction()?)
-            }
-            StorageBackend::S3(s3) => Box::new(s3.start_storage_transaction()),
-        };
-
-        let res = f(trans.as_mut())?;
-        trans.complete()?;
-        Ok(res)
-    }
-
     // Store all files in `root_dir` into the backend under `prefix`.
     //
     // This returns (map<filename, mime type>, set<compression algorithms>).
-    pub(crate) fn store_all(
+    pub(crate) async fn store_all(
         &self,
         prefix: &Path,
         root_dir: &Path,
     ) -> Result<(HashMap<PathBuf, String>, HashSet<CompressionAlgorithm>)> {
         let mut file_paths_and_mimes = HashMap::new();
         let mut algs = HashSet::with_capacity(1);
 
-        let blobs = get_file_list(root_dir)?
+        let blobs: Vec<_> = get_file_list(root_dir)?
```

Comment:
This whole expression should be in a `spawn_blocking` since it's reading from the filesystem and doing a lot of compression.